first draft for 30 by 30 content (#8)
coreyja/toys: 30by30/ideas.md

## 30 by 30 
Our very cool segment for posting tweets about:
- Tips
- Hacks  
- Do's & Don'ts!

This is a place where we can generate ideas!

## Tweet1
🔄 You can swap two values with creating a temp variable! 
#Rust30by30 #Rustc #Rustlang #Day1
```
    let a = 1;
    let b = 2;
    let (a, b) = (b, a);
    println!("{}, {}",a,b);
    //output: 2, 1
```

## Tweet2
🐞 Rust Hack: Use the `dbg!` macro to print variables and expressions during debugging. No more `println!`
#RustHacks #Debugging #Rust30by30 #Rustc #Rustlang #Day2

## Tweet 3
✅ DO: Use Rust's built-in testing framework to write unit tests and integration tests.
❌ DON'T: Skip testing, it's crucial for maintainable code! #RustTesting #TDD

> #Rust30by30 #Rustc #Rustlang #Day3

Reviewer: Lets include a quick example of a test with the built in framework, and maybe a link to the docs or something for people to learn more!

## Tweet 4
🧩 Traits are not a method overloading, but a way to define methods that multiple types can implement, enabling polymorphism!
#Rust30by30 #Rustc #Rustlang #Day4
```
trait Printable {
    fn print(&self);
}

struct NewsArticle {
    title: String,
}

impl Printable for NewsArticle {
    fn print(&self) {
        println!("News Article: {}", self.title);
    }
}

struct BlogPost {
    title: String,
}

impl Printable for BlogPost {
    fn print(&self) {
        println!("Blog Post: {}", self.title);
    }
}
```
## Tweet 5
⚠️ Rust Tip: Use `.unwrap()` cautiously!

> ✅ DO: Use it in tests or when you're absolutely sure a `Result/Option` is `Some/Ok`.

Reviewer: ```suggestion
✅ DO: Use it in tests or when you're absolutely sure a `Result/Option` is `Some/Ok` but you can't convey that in the type system.
```

❌ DON'T: Use in production code where errors are possible.
🔍 Better alternatives:
• ❓ `?` operator
• 🧩 `match`
• 🎯 `expect()`

📚 Read more: https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html
#RustErrorHandling #Rust30by30 #Rustc #Rustlang #Day5

## Tweet 6
🔧 Rust Trick

> Use the `#[derive(Debug, Clone)]` attribute to automatically generate implementations for the Debug and Clone traits for your struct!

Reviewer: Maybe a call back to the `dbg!` macro from the previous days!

And wondering if there is a quick thing we can say about _why_ you want `Clone`...

```
#[derive(Debug, Clone)]
struct Person {
    name: String,
    age: u32,
}
```
#RustTricks #Rust30by30 #Rustc #Rustlang #Day6

## Tweet 7
💡 Rust Tip #1: 
✅ DO: use `match` for pattern matching. 
❌ DON'T: rely on `if-else` for enum handling. 
#RustLang #30by30  #Rustc #Rustlang #Day7

```
enum TrafficLight {
    Red,
    Yellow,
    Green,
}

fn get_light_action(light: TrafficLight) -> &'static str {
    match light {
        TrafficLight::Red => "Stop",
        TrafficLight::Yellow => "Caution",
        TrafficLight::Green => "Go",
    }
}
```

## Tweet 8
⚡ Rust Tip: Speed up compile times by using `cargo check` instead of `cargo build` during development. 

> It skips building everything!

Reviewer: ```suggestion
It skips building everything but still checks your code for any compilation errors!
```

#RustTricks #Rust30by30 #Rustc #Rustlang #Day8

## Tweet 9
✅ DO: use `Option<T>` for values that might be absent.

> ❌ DON'T: use `null` pointers - Rust doesn't have them! 

Reviewer: I like talking about leaning into `Option<T>` but wonder if the Do/Don't is the best here since you like _cant_ do the don't version lol

```
struct animal {
    name: String,
    age: u32,
    fav_treats: Option<String>
}
```
#RustTricks #Rust30by30 #Rustc #Rustlang #Day9

## Tweet 10
✅ DO: use `&str` for string slices and String for owned strings. 

> ❌ DON'T: use `&String` as a function parameter. 

Reviewer: Ohh this one def deservers a _why_, but I'm not sure I have/know the simple quick tweet version. But we should create find one!

#RustTricks #Rust30by30 #Rustc #Rustlang #Day10
```
fn print_string(s: &str) {
    println!("{}", s);
}

fn main() {
    let my_string = String::from("Hello, world!");
    print_string(&my_string);
}
```

## Tweet 11
🎨 Rust Tip: Use `cargo fmt` to automatically format your code according to Rust style guidelines.

> #RustTips #CodeStyle #Rust30by30 #Rustc #Rustlang #Day11

Reviewer: Lets extend this to show people how to setup VS Code to automatically do this on save!

Which of course I don't have an easy link to, but there is a setting in VS Code that does this nicely!

## Tweet 12
🔍 Rust Tip: Use `cargo expand` to see what your macros are actually generating!
✅ DO: Leverage this tool for debugging and understanding macro expansions.
❌ DON'T: Guess what your macros are doing—get the actual output!

> #RustMacros #Rust30by30 #Rustc #Rustlang #Day12

Reviewer: Let's include a screenshot example of expanding some 'simple' macro like `println!`

## Tweet 13
🧪 Rust Hack: Use the `#[cfg(test)]` attribute to write inline tests in your source files.
#RustTesting #Rust30by30 #Rustc #Rustlang #Day13

## Tweet 14
🔍 Rust Tip: Use `cargo clippy` to catch common mistakes and improve your code quality.
#RustLinter #CodeQuality #Rust30by30 #Rustc #Rustlang #Day14

## Tweet 15
✅ DO: Use `Vec<T>` for growable arrays.
❌ DON'T: Use fixed-size arrays when you need dynamic sizing.

> #RustCollections #Rust30by30 #Rustc #Rustlang #Day15

Reviewer: This one it the first one I'm not in love with 🤷 

Feels obvious especially since it doesn't really provide an alternative or anything

## Tweet 16
🔄 Rust Hack: Use the `std::mem::swap()` function to efficiently swap values.

> #RustPerformance #Rust30by30 #Rustc #Rustlang #Day16

Reviewer: Maybe this should say to `swap mutable references`? Since Tip 2ish was about the in-place variable swap

Alternatively we may want to cut this one... Not sure I've ever used `std::mem::swap()`

## Tweet 17

> 🔍 Rust Tip: Use the `std::fmt` module to create custom formatting!

Reviewer: ```suggestion
🔍 Rust Tip: Use the `std::fmt::Display` trait to customize how your structs are printed!
```

#RustFormatting #Rust30by30 #Rustc #Rustlang #Day17
```
use std::fmt;
struct Person { name: String, age: u32 }
impl fmt::Display for Person {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {

>         write!(f, "{} ({})", self.name, self.age)

Reviewer: ```suggestion
        write!(f, "{} ({} years old)", self.name, self.age)
```

Just to make the name vs age more obvious

    }
}
```
## Tweet 18
⏱️ Rust Tip: Use `cargo bench` for benchmarking your code's performance.

> #RustPerformance #Benchmarking #Rust30by30 #Rustc #Rustlang #Day18

Reviewer: Oh lets include an example for this one, but I like it!

## Tweet 19
🔍 Rust Hack: Use SQLX for type-safe SQL queries in Rust!
✅ DO: Leverage compile-time checked queries with `sqlx::query!`.
❌ DON'T: Risk runtime errors with string-based queries.
#RustDatabase #SQLX #Rust30by30 #Rustc #Rustlang #Day19
```
let users = sqlx::query!("SELECT * FROM users").fetch_all(db).await?;

> ```

Reviewer: _Might_ want to remove this one for now, and save it for another series on crates

I think it might be nice to keep this series 'crate-less'

## Tweet 20
⚠️ Rust Hack: Use the `#[must_use]` attribute to ensure important return values aren't ignored.
#RustSafety #Rust30by30 #Rustc #Rustlang #Day20
```

> #[must_use]

Reviewer: Lets link out to this guide on when to use this!

https://std-dev-guide.rust-lang.org/policy/must-use.html

And maybe we can spice up the example to show why it might be nice to add to things

Reviewer: Unresolving cause I want to think about this example a bit more...

I like the tip alot! But using a `Result` here as the return type might make the example a bit less helpful cause the Result struct is marked `must_use` so not sure marking this function really help much 🤔 

Maybe we have the function return an `u32` and write a comment in the body saying how its expected to return `0` for success and `1` for failure, kinda like the exit status of programs in the CLI
Then it will make more sense to add `must_use` cause we want someone to look at the return value to tell if it failed, and Rust won't do that for us since we aren't returning a result

Author: something like this 
``` rust
#[must_use]
fn important_function() -> u32 {
    // Return 0 for success or 1 for failure
    return 0
}
```

fn important_function() -> Result<(), String> {
}
```

## Tweet 21
✅ DO: Use `std::collections::HashMap` for key-value storage.
❌ DON'T: Reinvent the wheel with custom data structures.

> #RustCollections #Rust30by30 #Rustc #Rustlang #Day21

Reviewer: I like the idea here, but maybe we choose some 'less' common data types here than `HashMap`?

Like `HashSet` might be a reasonable one... Maybe we can name a few less known structures that Rust has built in

## Tweet 22
📚 Rust Tip: Use `cargo doc` to generate documentation for your project.

> #RustDocs #Rust30by30 #Rustc #Rustlang #Day22

Reviewer: I think we should expand on this one a bit, maybe talk about how the `///` comments are doc comments that will show up in this documentation

## Tweet 23
📦 Rust Tip: Embrace external crates! They're a strength, not a weakness.
✅ DO: Leverage the ecosystem for well-tested solutions.
❌ DON'T: Reinvent the wheel when great crates already exist.

> #RustEcosystem #Crates #Rust30by30 #Rustc #Rustlang #Day23

Reviewer: Maybe this can be the last one in the series, and a way to lead into the next series about cool crates?

## Tweet 24
🔁 Rust Hack: Use the `std::iter` module for powerful iterator operations.

> #RustIterators #Rust30by30 #Rustc #Rustlang #Day24

Reviewer: Some examples here would probably be good! Like why are iterators powerful? What can we do with them?

## Tweet 25
🎨 Rust Tip: Use the `color_eyre` crate for customizable error reports!
✅ DO: Add by running `cargo add color_eyre` and then `cargo check` to see the magic!
❌ DON'T: Settle for plain, hard-to-read error messages.

> #RustErrorHandling #Rust30by30 #Rustc #Rustlang #Day25

Reviewer: Let's save it for the next series!

## Tweet 26
🔄 Rust Tip: Use `cargo update` to update your project's dependencies.

> #RustDependencies #Rust30by30 #Rustc #Rustlang #Day26

Reviewer: Any special cli options or anything that we might want to showcase?

This is a _bit_ basic on its own maybe 🤔

## Tweet 27
✅ DO: Use `std::sync::Arc` for thread-safe reference counting.
❌ DON'T: Use `Rc` in multi-threaded contexts.

> #RustConcurrency #Rust30by30 #Rustc #Rustlang #Day27

Reviewer: Probably good to explain why not

Here I might say something like `Atomic Reference Counting is safer in multi-threaded scenarios, and in almost all cases the atomic counting isn't meaningfully slower but does provide more safety!`

## Tweet 28
🏗️ Rust Hack: Use the `#[derive(Default)]` attribute for easy struct initialization.

> #RustProductivity #Rust30by30 #Rustc #Rustlang #Day28

Reviewer: Lets show an example! And maybe move this up in the list a bit

## Tweet 29
✅ DO: Leverage closures for flexible and reusable code. 💪
❌ DON'T: Avoid using closures when a simple function would suffice, as it can lead to unnecessary complexity.
```
fn add_one<F>(x: i32, f: F) -> i32
where F: Fn(i32) -> i32 {
    f(x) + 1
}

> ```

Reviewer: I'm... Not really sure what this example is trying to show 😆

Author: fixed it! Let me know what you think!

#RustCodeSnippets #Rust30by30 #Rustc #Rustlang #Day29

## Tweet 30
🌟 Rust Tip: Use the `axum` crate for building HTTP servers and handling requests easily!

> #RustCommunity #Rust30by30 #Rustc #Rustlang #Day30

Reviewer: For the next series too!

#Rust30by30 #Rustc #Rustlang #Day3

## Tweet 4
🧩 Traits are not a method overloading, but a way to define methods that multiple types can implement, enabling polymorphism!
```rust
trait Printable {
    fn print(&self);
}

struct NewsArticle {
    title: String,
}

impl Printable for NewsArticle {
    fn print(&self) {
        println!("News Article: {}", self.title);
    }
}

struct BlogPost {
    title: String,
}

impl Printable for BlogPost {
    fn print(&self) {
        println!("Blog Post: {}", self.title);
    }
}
```
#Rust30by30 #Rustc #Rustlang #Day4

## Tweet 5
⚠️ Rust Tip: Use `.unwrap()` cautiously!
✅ DO: Use it in tests or when you're absolutely sure a `Result/Option` is `Some/Ok` but you can't convey that in the type system.
❌ DON'T: Use in production code where errors are possible.
🔍 Better alternatives:
• ❓ `?` operator
• 🧩 `match`
• 🎯 `expect()`

📚 Read more: https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html
#RustErrorHandling #Rust30by30 #Rustc #Rustlang #Day5

## Tweet 6
🔧 Rust Trick
Use the `#[derive(Debug, Clone)]` attribute to automatically generate implementations for the Debug and Clone traits for your struct!

> 💡 `Clone` is useful when you need multiple ownership of data!

Reviewer: I like this Lightbulb call out! Lets add one for `Debug` too, about it being useful for getting a developer friendly output for your struct

```rust
#[derive(Debug, Clone)]
struct Person {
    name: String,
    age: u32,
}

fn main() {
    let alice = Person { name: "Alice".into(), age: 30 };

> dbg!(alice.clone()); // creates a copy of the struct and prints it

Reviewer: SuperNit: Indentation here is off I think!

    age: u32,
    fav_treats: Option<String>
}
```
#RustTricks #Rust30by30 #Rustc #Rustlang #Day9

## Tweet 10

🏗️ Rust Hack: Use the `#[derive(Default)]` attribute for easy struct initialization.

example:
```rust
#[derive(Default)]
struct Person {
    name: String,
    age: u32,
}

fn main() {
    let person = Person::default();
    println!("{}", person.name);
}
```

You can use `#[derive(Default)]` for custom structs as well!
```rust
#[derive(Default)]
struct Person {
    name: String,
    age: u32,
}

fn main() {
    let custom_person = Person {
        name: "rust biggest fan".to_string(),
        age: 32,
        ..Default::default()
    };
    println!("Custom Person: {:?}", custom_person.name);
}
```

#RustTricks #Rust30by30 #Rustc #Rustlang #Day10

## Tweet 11
🎨 Rust Tip: Use `cargo fmt` to automatically format your code according to Rust style guidelines.
📝 VS Code Setup for Auto-Format on Save:
1. Install the Rust Analyzer extension
2. Open settings
3. Search for "rust format"
4. Enable "Format on Save"

or you can add the following to your settings.json:
```json
   "[rust]": {
    "editor.defaultFormatter": "rust-lang.rust-analyzer",
    "editor.formatOnSave": true

> }

Reviewer: Thank you for including the JSON as well as instructions!

Love that cause I want the JSON 100% of the time, but not everyone does lol

```

Link to ref: https://users.rust-lang.org/t/how-to-active-format-on-save-with-rust-analyzer-for-vs-code/41701

#RustTips #CodeStyle #Rust30by30 #Rustc #Rustlang #Day11

## Tweet 12
⚡ Rust Tip: `cargo expand` - Your X-ray Vision for Macros!

It shows you what's happening behind the scenes when macros are expanded.

example:
```rust
macro_rules! create_function {
    ($func_name:ident) => {
        fn $func_name() {
            println!("You called {:?}()", stringify!($func_name));
        }
    };

> }

Reviewer: Oh this is fun! I wasn't thinking of writing our own macro here. BUT I really like it so 💯 !

create_function!(foo);
create_function!(bar);

fn main() {
    foo();
    bar();
}
```

run `cargo expand` to see the output!

output:
```rust
// notice how all the imports are being unveiled!
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;

// notice how the function is being unveiled!
fn foo() {
    {
        ::std::io::_print(format_args!("You called {0:?}()\n", "foo"));
    };
}
fn bar() {
    {
        ::std::io::_print(format_args!("You called {0:?}()\n", "bar"));
    };
}
fn main() {
    foo();
    bar();
}
```

After running `cargo expand`, you can see:
- All the imports
- Macro expansion
- Better code readability
- Learn more about Rust internals

Remember what you see is what the compiler sees! 🚀

Learn more here: https://docs.rs/crate/cargo-expand/0.1.3/source/README.md

#RustMacros #Rust30by30 #Rustc #Rustlang #Day12

## Tweet 13
🔍 Rust Tip: Use `cargo clippy` to catch common mistakes and improve your code quality.

> #RustLinter #CodeQuality #Rust30by30 #Rustc #Rustlang #Day13

Reviewer: Want to expand the wording here, but don't have a specific thought and want to keep running down this list so comment to come back to it later!

## Tweet 14

> 🧪 Rust Tip: Use `#[test]` with `#[ignore]` for long-running or resource-intensive tests.

Reviewer: Oh this is a new one right? Love it!

Author: Yes indeed it is!

```rust
#[cfg(test)]
mod tests {
    use std::thread;
    use std::time::Duration;

    #[test]
    fn quick_test() {
        assert_eq!(2 + 2, 4);
    }

    #[test]
    #[ignore]
    fn slow_test() {
        // Simulate a long-running test
        thread::sleep(Duration::from_secs(5));
        assert!(true);
    }
}
```

Run all tests: `cargo test`
Run ignored tests: `cargo test -- --ignored`

This helps keep your test suite fast while still allowing thorough testing when needed!

#RustTesting #TestOptimization #Rust30by30 #Rustc #Rustlang #Day14

## Tweet 15
🔍 Rust Tip: Use `Vec::with_capacity(n)` when you know the approximate size of your vector beforehand.

```rust
let mut vec = Vec::with_capacity(10);
for i in 0..10 {
    vec.push(i);
}
```
#RustCollections #Rust30by30 #Rustc #Rustlang #Day15

## Tweet 16

> 🔍 Rust Tip: Use `std::mem::size_of::<T>()` to check the size of types at compile-time.

Reviewer: This is a good one too! Love it!

This is really good for the 'low level' system type programming that Rust in known for!

```rust
let size = std::mem::size_of::<i32>();
println!("Size of i32: {}", size);
```
#RustPerformance #Rust30by30 #Rustc #Rustlang #Day16

## Tweet 17
🔍  Rust Tip: Use the `std::fmt::Display` trait to customize how your structs are printed!
```
use std::fmt;
struct Person { name: String, age: u32 }
impl fmt::Display for Person {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} ({} years old)", self.name, self.age)
    }
}
```
#RustFormatting #Rust30by30 #Rustc #Rustlang #Day17

## Tweet 18
⏱️ Rust Tip: Use `cargo bench` for benchmarking your code's performance.

```rust
#![feature(test)]

extern crate test;

pub fn fibonacci(n: u64) -> u64 {
    match n {
        0 => 0,
        1 => 1,
        _ => fibonacci(n - 1) + fibonacci(n - 2),
    }

> }

Reviewer: Love a good Fib impl!

This suggestion is mostly just cause I love Fib (I was actually explaining it and recursion to Brandi last night just cause the word fibonacci came up in a show lol)

Might be nice to have an `iterative_fib` function that doesn't use recursion and does it 'iteratively' instead and then benchmark can compare the two! (Iterative should be much faster than this basic recursive version!)

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;

    #[bench]
    fn bench_fib_10(b: &mut Bencher) {
        b.iter(|| fibonacci(10));
    }

    #[bench]
    fn bench_fib_20(b: &mut Bencher) {
        b.iter(|| fibonacci(20));
    }
}

fn main() {
    println!("Hello, world!");
}

run `cargo bench` to see the output!

```
output:
```
running 2 tests
test tests::bench_fib_10 ... bench:           0.51 ns/iter (+/- 0.01)
test tests::bench_fib_20 ... bench:           0.51 ns/iter (+/- 0.01)
```

The output from `cargo bench` shows the results of running two benchmark tests:

1. `bench_fib_10`: This test took approximately 0.51 nanoseconds per iteration.
2. `bench_fib_20`: This test also took about 0.51 nanoseconds per iteration.

> The "+/- 0.01" indicates the margin of error for each measurement. These results suggest that both `fibonacci(10)` and `fibonacci(20)` are executing extremely quickly, with no significant difference in performance between them. This is unexpected and might indicate an optimization by the compiler, such as constant folding, which is computing the results at compile-time rather than runtime.

Reviewer: I think 10 to 20 just isn't a big enough difference to tell likely...

I'd compare 100 to like 1000 maybe and hope to see a bigger difference! `fib(20)` is just too 'easy' for a computer to do 😆

#RustPerformance #Benchmarking #Rust30by30 #Rustc #Rustlang #Day18

## Tweet 19
⚠️ Rust Hack: Use the `#[must_use]` attribute to ensure important return values aren't ignored.

link to ref: https://std-dev-guide.rust-lang.org/policy/must-use.html
```
#[must_use]
fn important_function() -> Result<(), String> {
}
```
#RustSafety #Rust30by30 #Rustc #Rustlang #Day19

## Tweet 20
✅ DO: Use `std::collections::HashSet` for storing unique values.
❌ DON'T: Reinvent the wheel with custom data structures.

> #RustCollections #Rust30by30 #Rustc #Rustlang #Day20

Reviewer: This one might be fun to add a little 'Sidenote' or whatever

About how a `HashSet` is implemented as a `HashMap` where the value is the empty tuple!

So `HashSet<String>` is basically the same as `HashMap<String, ()>`!

## Tweet 21
📚 Rust Tip: Use `cargo doc` to generate documentation for your project.

Doc comments use three slashes `///` and support Markdown
They appear in the generated documentation
Regular double line comments `//` won't be included in the documentation.

example:
Run `cargo doc --open` to generate and view your documentation!

```rust
/// The `add` function adds two numbers
/// 
/// # Examples
/// 
/// let result = add(2, 3);
/// assert_eq!(result, 5);
/// 
pub fn add(a: i32, b: i32) -> i32 {
    a + b
}
```
#RustDocs #DocComments #Rust30by30 #Rustc #Rustlang #Day21

## Tweet 22

> 🔒 Rust Tip: Use the `#[non_exhaustive]` attribute for future-proofing your enums and structs.

Reviewer: Oh love this one!

I think just a bit more explanation would be great! Something about how `non_exhaustive` forces people to have a `_` pattern in their matches that will match any new items you add later.

This is basically what you have but a bit more explicit so the reader doesn't have to read between the lines at all

```rust
#[non_exhaustive]
pub enum Color {
    Red,
    Green,
    Blue,
}

fn main() {
    let color = Color::Red;
    match color {
        Color::Red => println!("It's red!"),
        Color::Green => println!("It's green!"),
        Color::Blue => println!("It's blue!"),
        _ => println!("It's a new color!"),
    }
}
```

This allows you to add new variants later without breaking existing code.

#RustPatterns #APIDesign #Rust30by30 #Rustc #Rustlang #Day22

## Tweet 23

> 🔁 Rust Hack: Use the `std::iter` module for powerful iterator operations.

Reviewer: I love iters so I think there is something here, but this one just feels a bit 'weak' to me right now...

Will come back and try to beef it up!

```rust
let nums = vec![1, 2, 3, 4, 5];
let sum: i32 = nums.iter().sum();
println!("Sum: {}", sum);
```
#RustIterators #Rust30by30 #Rustc #Rustlang #Day23

## Tweet 24

> 💡 Rust tip: Use `#[allow(unused_variables)]` to suppress warnings for unused variables.

Reviewer: I wonder if I actually want to give this tip... On one hand its good to know, but on the other I don't think people should be ignoring warnings 😆 

But I think it is good to know!
And maybe we generalize it a tiny bit and say that it can be used to suppress any warning (including those from clippy)

```rust
#[allow(unused_variables)]
fn print_x() {
    let x = 5;
    let y = 10;

>     // ooups we forgot to use y!

Reviewer: SuperNit: I'd remove this comment since if we added the `allow` it must be 'on purpose' so probably not an `oops` lol

    println!("x: {}", x);
}

fn main() {
    print_x();
}
```

## Tweet 25
🔄 Rust Tip: Use `cargo update` to update your project's dependencies.

example usuage:
1) Update all: `cargo update`
2) Update specific: `cargo update -p <package-name>`

> 3) Update all dev-dependencies: `cargo update -d`

Reviewer: Love these examples!

One good thing to add might be a sentence about how Cargo will do 'semantic version' upgrades. So if you asked for `1` in your Cargo.toml it will upgrade to any `1.x` version but NOT a `2.0` since that might have breaking changes

#RustDependencies #Rust30by30 #Rustc #Rustlang #Day25

## Tweet 26
✅ DO: Use `std::sync::Arc` for thread-safe reference counting.
❌ DON'T: Use `Rc` in multi-threaded contexts.

Atomic Reference Counting is safer in multi-threaded scenarios, and in almost all cases the atomic counting isn't meaningfully slower but does provide more safety!

#RustConcurrency #Rust30by30 #Rustc #Rustlang #Day26

## Tweet 27
✅ DO: use `&str` for string slices and String for owned strings. 
❌ DON'T: use `&String` as a function parameter. 

Why? Using `&str` is more flexible:

> 1. It accepts both `&str` and `String` as arguments.

Reviewer: Nit: Might say it accepts both `&str` and `&String` cause you do still need the borrow to call a function with `&str`

2. It's more efficient as it doesn't need to allocate memory for the string.
3. It's more flexible as it can be used to pass string literals and string slices.
4. It clearly communicates that the function won't modify the string.
```
fn print_string(s: &str) {
    println!("{}", s);
}

fn main() {
    let my_string = String::from("Hello, world!");
    print_string(&my_string);
}
```
#RustProductivity #Rust30by30 #Rustc #Rustlang #Day27

## Tweet 28
✅ DO: Leverage closures for flexible and reusable code. 💪

> ❌ DON'T: Avoid using closures when a simple function would suffice, as it can lead to unnecessary complexity.

Reviewer: Still not sure about this one...

Do you think this is relevant coming from a more JS/TS background? Just asking cause in like Ruby or Rust I don't lean to closures much (more true in Rust than Ruby I guess)

So this tip feels a bit weird to me cause I wouldn't have ever thought to do the DONT version lol

But in JS/TS (and def React land) closures are more common/needed. So maybe there is value in this tip!

Author: Make sense, I agree

> ```

Reviewer: This example is kinda a `counter example` right? Like its doing the DONT version?

Lets call that out explicitly!

fn apply_operation<F>(x: i32, y: i32, operation: F) -> i32
where
    F: Fn(i32, i32) -> i32,
{
    operation(x, y)
}

fn main() {
    // Using a closure for addition
    let result_add = apply_operation(5, 3, |a, b| a + b);
    println!("5 + 3 = {}", result_add);

    // Using a closure for multiplication
    let result_multiply = apply_operation(5, 3, |a, b| a * b);
    println!("5 * 3 = {}", result_multiply);

    // Using a more complex closure
    let result_custom = apply_operation(5, 3, |a, b| (a + b) * (a - b));
    println!("(5 + 3) * (5 - 3) = {}", result_custom);
}
```
#RustCodeSnippets #Rust30by30 #Rustc #Rustlang #Day28

## Tweet 29

> 🤝 Rust Tip: Contribute to the Rust community!

Reviewer: ❤️ ❤️ ❤️

✅ DO:
- Report bugs you encounter
- Help answer questions on forums like users.rust-lang.org
- Contribute to open-source Rust projects
- Write blog posts or tutorials about your Rust experiences

❌ DON'T:
- Be afraid to ask for help
- Hesitate to share your knowledge, no matter your skill level

Remember: Every contribution, big or small, helps the Rust ecosystem grow! 🦀

#RustCommunity #OpenSource #Rust30by30 #Rustc #Rustlang #Day29

## Tweet 30

> 📦 Rust Tip: Embrace external crates! They're a strength, not a weakness.

Reviewer: Love it!!

Might be fun to add a teaser here to an upcoming series on crates!
